File: src/download.rs
use std::{fs, io::Write};

use arxiv::{Arxiv, ArxivQueryBuilder};
use serde::{Deserialize, Serialize};

/// Sanitize a filename to be Windows-compatible
fn sanitize_filename(name: &str) -> String {
    // Replace invalid Windows filename characters with underscores
    let invalid_chars = ['<', '>', ':', '"', '/', '\\', '|', '?', '*'];
    let mut sanitized = name.to_string();
    for ch in invalid_chars {
        sanitized = sanitized.replace(ch, "_");
    }
    // Trim leading/trailing whitespace and dots
    sanitized = sanitized.trim().trim_end_matches('.').to_string();
    // Limit filename length to 200 characters to be safe
    if sanitized.len() > 200 {
        sanitized.truncate(200);
    }
    sanitized
}

const JSON_FILE: &str = "metadata.jsonl";
const PDF_DIRECTORY: &str = "pdfs/";
const TEXT_DIRECTORY: &str = "texts/";

#[derive(Serialize, Deserialize, Clone)]
pub struct SerDesArxiv {
    pub id: String,
    pub updated: String,
    pub published: String,
    pub title: String,
    #[serde(skip_serializing)]
    pub summary: String,
    pub authors: Vec<String>,
    pub primary_category: String,
    pub categories: Vec<String>,
    pub pdf_url: String,
    pub html_url: String,
    pub comment: Option<String>,
}

impl SerDesArxiv {
    fn from_arxiv(arxiv_paper: Arxiv) -> Self {
        Self {
            id: arxiv_paper.id,
            updated: arxiv_paper.updated,
            published: arxiv_paper.published,
            title: arxiv_paper.title,
            summary: arxiv_paper.summary,
            authors: arxiv_paper.authors,
            primary_category: arxiv_paper.primary_category,
            categories: arxiv_paper.categories,
            pdf_url: arxiv_paper.pdf_url.replace("httpss", "https"),
            html_url: arxiv_paper.html_url.replace("httpss", "https"),
            comment: arxiv_paper.comment,
        }
    }

    pub async fn fetch_pdf(&self, out_path: &str) -> anyhow::Result<()> {
        let body = reqwest::get(&self.pdf_url).await?.bytes().await?;
        let out_path = if out_path.ends_with(".pdf") {
            out_path.to_string()
        } else {
            format!("{}.pdf", out_path)
        };
        let mut file = fs::File::create(out_path)?;
        file.write_all(&body)?;
        Ok(())
    }

    // TODO: make this function actually usable
    // pub async fn fetch_text(&self, out_path: &str) -> anyhow::Result<()> {
    //     let body = reqwest::get(&self.html_url).await?.bytes().await?;
    //     let html_text = from_read(&body[..], 20)?;
    //     let out_path = if out_path.ends_with(".txt") {
    //         out_path.to_string()
    //     } else {
    //         format!("{}.txt", out_path)
    //     };
    //     fs::write(out_path, &html_text)?;
    //     Ok(())
    // }

    pub fn write_summary(&self, out_path: &str) -> anyhow::Result<()> {
        let out_path = if out_path.ends_with(".txt") {
            out_path.to_string()
        } else {
            format!("{}.txt", out_path)
        };
        let summary = self.summary.clone();
        fs::write(out_path, summary)?;
        Ok(())
    }
}

pub async fn download_arxiv_papers(
    search_query: String,
    num_results: i32,
    save_metadata: bool,
    save_pdfs: bool,
    save_summaries: bool,
) -> anyhow::Result<()> {
    let query = ArxivQueryBuilder::new()
        .search_query(&search_query)
        .start(0)
        .max_results(num_results)
        .sort_by("submittedDate")
        .sort_order("descending")
        .build();
    let arxivs = arxiv::fetch_arxivs(query).await?;
    let mut jsonl_text: String = "".to_string();
    for a in arxivs {
        let paper = SerDesArxiv::from_arxiv(a);
        if save_metadata {
            let paper_copy = paper.clone();
            let paper_metadata = serde_json::to_string(&paper_copy)?;
            jsonl_text += &format!("{}\n", paper_metadata);
        }
        if save_pdfs {
            let pdf_dir_exists = fs::exists(PDF_DIRECTORY)?;
            if !pdf_dir_exists {
                fs::create_dir(PDF_DIRECTORY)?;
            }
            let sanitized_title = sanitize_filename(&paper.title);
            let path = format!("{}/{}", PDF_DIRECTORY, sanitized_title);
            paper.fetch_pdf(&path).await?;
        }
        if save_summaries {
            let txt_dir_exists = fs::exists(TEXT_DIRECTORY)?;
            if !txt_dir_exists {
                fs::create_dir(TEXT_DIRECTORY)?;
            }
            let sanitized_title = sanitize_filename(&paper.title);
            let path = format!("{}/{}.txt", TEXT_DIRECTORY, sanitized_title);
            paper.write_summary(&path)?;
        }
    }
    if !jsonl_text.is_empty() {
        fs::write(JSON_FILE, &jsonl_text)?;
    }
    Ok(())
}

#[cfg(test)]
mod test {
    use super::*;
    use serial_test::serial;
    use std::path::Path;

    #[tokio::test]
    #[serial]
    async fn integration_test_defaults() {
        if Path::new(PDF_DIRECTORY).exists() {
            fs::remove_dir_all(PDF_DIRECTORY).expect("Should be able to remove PDF directory");
        }
        if Path::new(TEXT_DIRECTORY).exists() {
            fs::remove_dir_all(TEXT_DIRECTORY).expect("Should be able to remove text directory");
        }
        if Path::new(JSON_FILE).exists() {
            fs::remove_file(JSON_FILE).expect("Should be able to remove metadata.jsonl file");
        }
        let result = download_arxiv_papers("cs.CL".to_string(), 5, true, false, false).await;
        match result {
            Ok(_) => {}
            Err(e) => {
                eprintln!("An error occurred: {}", e.to_string());
                assert!(false)
            }
        }
        let file_exists = fs::exists(JSON_FILE)
            .expect("Should be able to check the existance of the metadata.jsonl file");
        assert!(file_exists);
        let content =
            fs::read_to_string(JSON_FILE).expect("Should be able to read metadata.jsonl file");
        assert!(content.len() > 0);
    }

    #[tokio::test]
    #[serial]
    async fn integration_test_pdfs() {
        if Path::new(PDF_DIRECTORY).exists() {
            fs::remove_dir_all(PDF_DIRECTORY).expect("Should be able to remove PDF directory");
        }
        if Path::new(TEXT_DIRECTORY).exists() {
            fs::remove_dir_all(TEXT_DIRECTORY).expect("Should be able to remove text directory");
        }
        if Path::new(JSON_FILE).exists() {
            fs::remove_file(JSON_FILE).expect("Should be able to remove metadata.jsonl file");
        }
        let result = download_arxiv_papers("cs.CL".to_string(), 2, false, true, false).await;
        match result {
            Ok(_) => {}
            Err(e) => {
                eprintln!("An error occurred: {}", e.to_string());
                assert!(false)
            }
        }
        let file_exists = fs::exists(JSON_FILE)
            .expect("Should be able to check the existance of the metadata.jsonl file");
        assert!(!file_exists);
        let dir_exists: bool = fs::exists(PDF_DIRECTORY)
            .expect("Should be able to check the existance of the PDF directory");
        assert!(dir_exists);
        let dir_content =
            fs::read_dir(PDF_DIRECTORY).expect("Should be able to read the PDF directory");
        let mut count = 0;
        for entry in dir_content {
            let _dir_entry = entry.expect("Should be able to read entry");
            count += 1;
        }
        assert_eq!(count, 2);
    }

    #[tokio::test]
    #[serial]
    async fn integration_test_texts() {
        if Path::new(PDF_DIRECTORY).exists() {
            fs::remove_dir_all(PDF_DIRECTORY).expect("Should be able to remove PDF directory");
        }
        if Path::new(TEXT_DIRECTORY).exists() {
            fs::remove_dir_all(TEXT_DIRECTORY).expect("Should be able to remove text directory");
        }
        if Path::new(JSON_FILE).exists() {
            fs::remove_file(JSON_FILE).expect("Should be able to remove metadata.jsonl file");
        }
        let result = download_arxiv_papers("cs.CL".to_string(), 2, false, false, true).await;
        match result {
            Ok(_) => {}
            Err(e) => {
                eprintln!("An error occurred: {}", e.to_string());
                assert!(false)
            }
        }
        let file_exists = fs::exists(JSON_FILE)
            .expect("Should be able to check the existance of the metadata.jsonl file");
        assert!(!file_exists);
        let pdf_dir_exists: bool = fs::exists(PDF_DIRECTORY)
            .expect("Should be able to check the existance of the PDF directory");
        assert!(!pdf_dir_exists);
        let text_dir_exists: bool = fs::exists(TEXT_DIRECTORY)
            .expect("Should be able to check the existance of the text directory");
        assert!(text_dir_exists);
        let dir_content =
            fs::read_dir(TEXT_DIRECTORY).expect("Should be able to read the PDF directory");
        let mut count = 0;
        for entry in dir_content {
            let _dir_entry = entry.expect("Should be able to read entry");
            count += 1;
        }
        assert_eq!(count, 2);
    }

    #[tokio::test]
    #[serial]
    async fn integration_test_all() {
        if Path::new(PDF_DIRECTORY).exists() {
            fs::remove_dir_all(PDF_DIRECTORY).expect("Should be able to remove PDF directory");
        }
        if Path::new(TEXT_DIRECTORY).exists() {
            fs::remove_dir_all(TEXT_DIRECTORY).expect("Should be able to remove text directory");
        }
        if Path::new(JSON_FILE).exists() {
            fs::remove_file(JSON_FILE).expect("Should be able to remove metadata.jsonl file");
        }
        let result = download_arxiv_papers("cs.CL".to_string(), 2, true, true, true).await;
        match result {
            Ok(_) => {}
            Err(e) => {
                eprintln!("An error occurred: {}", e.to_string());
                assert!(false)
            }
        }
        let file_exists = fs::exists(JSON_FILE)
            .expect("Should be able to check the existance of the metadata.jsonl file");
        assert!(file_exists);
        let content =
            fs::read_to_string(JSON_FILE).expect("Should be able to read metadata.jsonl file");
        assert!(content.len() > 0);
        let pdf_dir_exists: bool = fs::exists(PDF_DIRECTORY)
            .expect("Should be able to check the existance of the PDF directory");
        assert!(pdf_dir_exists);
        let pdf_dir_content =
            fs::read_dir(PDF_DIRECTORY).expect("Should be able to read the PDF directory");
        let mut pdf_count = 0;
        for entry in pdf_dir_content {
            let _dir_entry = entry.expect("Should be able to read entry");
            pdf_count += 1;
        }
        assert_eq!(pdf_count, 2);
        let text_dir_exists: bool = fs::exists(TEXT_DIRECTORY)
            .expect("Should be able to check the existance of the text directory");
        assert!(text_dir_exists);
        let dir_content =
            fs::read_dir(TEXT_DIRECTORY).expect("Should be able to read the PDF directory");
        let mut count = 0;
        for entry in dir_content {
            let _dir_entry = entry.expect("Should be able to read entry");
            count += 1;
        }
        assert_eq!(count, 2);
    }

    #[test]
    fn test_serdes_arxiv_write_summary() {
        let paper = SerDesArxiv {
            id: "".to_string(),
            updated: "".to_string(),
            published: "".to_string(),
            title: "test_title".to_string(),
            summary: "This is a test summary.".to_string(),
            authors: vec![],
            primary_category: "".to_string(),
            categories: vec![],
            pdf_url: "".to_string(),
            html_url: "".to_string(),
            comment: None,
        };
        let out_path = "test_summary.txt";
        paper
            .write_summary(out_path)
            .expect("Should write summary to file");
        let written = fs::read_to_string(out_path).expect("Should read summary file");
        assert_eq!(written, "This is a test summary.");
        fs::remove_file(out_path).expect("Should clean up summary file");
    }

    #[test]
    fn test_serdes_arxiv_to_string() {
        let paper = SerDesArxiv {
            id: "".to_string(),
            updated: "".to_string(),
            published: "".to_string(),
            title: "test_title".to_string(),
            summary: "This is a test summary.".to_string(),
            authors: vec![],
            primary_category: "".to_string(),
            categories: vec![],
            pdf_url: "".to_string(),
            html_url: "".to_string(),
            comment: None,
        };
        let json_content = serde_json::to_string(&paper).expect("Should be able to serialize");
        assert!(!json_content.contains("summary"));
    }
}

File: internal/download/download.go
package download

import (
	"context"
	"encoding/json"
	"encoding/xml"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"strings"
	"time"
)

const (
	JSONFile     = "metadata.jsonl"
	PDFDirectory = "pdfs/"
	TextDirectory = "texts/"
	arxivAPIBase = "http://export.arxiv.org/api/query"
)

type ArxivPaper struct {
	ID              string   `json:"id"`
	Updated         string   `json:"updated"`
	Published       string   `json:"published"`
	Title           string   `json:"title"`
	Summary         string   `json:"-"` // skip in JSON like Rust
	Authors         []string `json:"authors"`
	PrimaryCategory string   `json:"primary_category"`
	Categories      []string `json:"categories"`
	PDFURL          string   `json:"pdf_url"`
	HTMLURL         string   `json:"html_url"`
	Comment         *string  `json:"comment,omitempty"`
}

// Atom XML structures for parsing arXiv API response
type Feed struct {
	XMLName xml.Name `xml:"feed"`
	Entries []Entry  `xml:"entry"`
}

type Entry struct {
	XMLName    xml.Name   `xml:"entry"`
	ID         string     `xml:"id"`
	Updated    string     `xml:"updated"`
	Published  string     `xml:"published"`
	Title      string     `xml:"title"`
	Summary    string     `xml:"summary"`
	Authors    []Author   `xml:"author"`
	Links      []Link     `xml:"link"`
	Categories []Category `xml:"category"`
	Comment    Comment    `xml:"http://arxiv.org/schemas/atom comment"`
}

type Comment struct {
	XMLName xml.Name `xml:"http://arxiv.org/schemas/atom comment"`
	Value   string   `xml:",chardata"`
}

type Author struct {
	Name string `xml:"name"`
}

type Link struct {
	Type  string `xml:"type,attr"`
	HRef  string `xml:"href,attr"`
	Rel   string `xml:"rel,attr"`
	Title string `xml:"title,attr"`
}

type Category struct {
	Term string `xml:"term,attr"`
}

func sanitizeFilename(name string) string {
	invalidChars := []rune{'<', '>', ':', '"', '/', '\\', '|', '?', '*'}
	sanitized := name
	for _, ch := range invalidChars {
		sanitized = strings.ReplaceAll(sanitized, string(ch), "_")
	}
	sanitized = strings.TrimSpace(sanitized)
	sanitized = strings.TrimRight(sanitized, ".")
	if len(sanitized) > 200 {
		sanitized = sanitized[:200]
	}
	return sanitized
}

func (p *ArxivPaper) FetchPDF(ctx context.Context, outPath string) error {
	client := &http.Client{
		Timeout: 30 * time.Second,
	}

	req, err := http.NewRequestWithContext(ctx, "GET", p.PDFURL, nil)
	if err != nil {
		return fmt.Errorf("failed to create request: %w", err)
	}

	resp, err := client.Do(req)
	if err != nil {
		return fmt.Errorf("failed to fetch PDF: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("failed to fetch PDF: HTTP %d", resp.StatusCode)
	}

	if !strings.HasSuffix(outPath, ".pdf") {
		outPath += ".pdf"
	}

	file, err := os.Create(outPath)
	if err != nil {
		return fmt.Errorf("failed to create file: %w", err)
	}
	defer file.Close()

	_, err = io.Copy(file, resp.Body)
	if err != nil {
		return fmt.Errorf("failed to write PDF: %w", err)
	}

	return nil
}

func (p *ArxivPaper) WriteSummary(outPath string) error {
	if !strings.HasSuffix(outPath, ".txt") {
		outPath += ".txt"
	}
	return os.WriteFile(outPath, []byte(p.Summary), 0644)
}

func fetchArxivPapers(ctx context.Context, searchQuery string, numResults int) ([]ArxivPaper, error) {
	baseURL, err := url.Parse(arxivAPIBase)
	if err != nil {
		return nil, fmt.Errorf("failed to parse base URL: %w", err)
	}

	params := url.Values{}
	params.Set("search_query", searchQuery)
	params.Set("start", "0")
	params.Set("max_results", fmt.Sprintf("%d", numResults))
	params.Set("sortBy", "submittedDate")
	params.Set("sortOrder", "descending")
	baseURL.RawQuery = params.Encode()

	client := &http.Client{
		Timeout: 30 * time.Second,
	}

	req, err := http.NewRequestWithContext(ctx, "GET", baseURL.String(), nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to fetch from arXiv API: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("arXiv API returned HTTP %d", resp.StatusCode)
	}

	var feed Feed
	decoder := xml.NewDecoder(resp.Body)
	if err := decoder.Decode(&feed); err != nil {
		return nil, fmt.Errorf("failed to parse XML response: %w", err)
	}

	papers := make([]ArxivPaper, 0, len(feed.Entries))
	for _, entry := range feed.Entries {
		paper := ArxivPaper{
			ID:              entry.ID,
			Updated:         entry.Updated,
			Published:       entry.Published,
			Title:           strings.TrimSpace(entry.Title),
			Summary:         strings.TrimSpace(entry.Summary),
			Authors:         make([]string, 0, len(entry.Authors)),
			PrimaryCategory: "",
			Categories:      make([]string, 0, len(entry.Categories)),
			Comment:         nil,
		}

		for _, author := range entry.Authors {
			paper.Authors = append(paper.Authors, author.Name)
		}

		for _, category := range entry.Categories {
			paper.Categories = append(paper.Categories, category.Term)
			if paper.PrimaryCategory == "" {
				paper.PrimaryCategory = category.Term
			}
		}

		for _, link := range entry.Links {
			if link.Rel == "alternate" && link.Type == "text/html" {
				paper.HTMLURL = strings.ReplaceAll(link.HRef, "httpss", "https")
			} else if link.Title == "pdf" {
				paper.PDFURL = strings.ReplaceAll(link.HRef, "httpss", "https")
			} else if link.Type == "application/pdf" {
				paper.PDFURL = strings.ReplaceAll(link.HRef, "httpss", "https")
			}
		}

		if entry.Comment.Value != "" {
			comment := entry.Comment.Value
			paper.Comment = &comment
		}

		papers = append(papers, paper)
	}

	return papers, nil
}

func DownloadArxivPapers(ctx context.Context, searchQuery string, numResults int, saveMetadata, savePDFs, saveSummaries bool) error {
	papers, err := fetchArxivPapers(ctx, searchQuery, numResults)
	if err != nil {
		return fmt.Errorf("failed to fetch papers: %w", err)
	}

	var jsonlLines []string

	for _, paper := range papers {
		if saveMetadata {
			paperCopy := paper
			metadataJSON, err := json.Marshal(paperCopy)
			if err != nil {
				return fmt.Errorf("failed to marshal metadata: %w", err)
			}
			jsonlLines = append(jsonlLines, string(metadataJSON))
		}

		if savePDFs {
			if err := os.MkdirAll(PDFDirectory, 0755); err != nil {
				return fmt.Errorf("failed to create PDF directory: %w", err)
			}
			sanitizedTitle := sanitizeFilename(paper.Title)
			path := filepath.Join(PDFDirectory, sanitizedTitle)
			if err := paper.FetchPDF(ctx, path); err != nil {
				return fmt.Errorf("failed to fetch PDF for %s: %w", paper.Title, err)
			}
		}

		if saveSummaries {
			if err := os.MkdirAll(TextDirectory, 0755); err != nil {
				return fmt.Errorf("failed to create text directory: %w", err)
			}
			sanitizedTitle := sanitizeFilename(paper.Title)
			path := filepath.Join(TextDirectory, sanitizedTitle+".txt")
			if err := paper.WriteSummary(path); err != nil {
				return fmt.Errorf("failed to write summary for %s: %w", paper.Title, err)
			}
		}
	}

	if len(jsonlLines) > 0 {
		content := strings.Join(jsonlLines, "\n") + "\n"
		if err := os.WriteFile(JSONFile, []byte(content), 0644); err != nil {
			return fmt.Errorf("failed to write metadata file: %w", err)
		}
	}

	return nil
}
